File: code/packages/go/starlark-compiler/starlarkcompiler_test.go
package starlarkcompiler

import "testing"

func TestOpValues(t *testing.T) {
	tests := map[Op]byte{
		LoadConst:     0x01,
		Add:           0x20,
		CmpEq:         0x30,
		Jump:          0x40,
		MakeFunction:  0x50,
		BuildList:     0x60,
		LoadSubscript: 0x70,
		GetIter:       0x80,
		LoadModule:    0x90,
		Print:         0xA0,
		Halt:          0xFF,
	}

	for op, want := range tests {
		if got := op.Byte(); got != want {
			t.Fatalf("%s byte = %#x, want %#x", op, got, want)
		}
	}
}

func TestOpFromByte(t *testing.T) {
	for _, op := range AllOps() {
		got, ok := OpFromByte(op.Byte())
		if !ok {
			t.Fatalf("OpFromByte(%#x) not found", op.Byte())
		}
		if got != op {
			t.Fatalf("OpFromByte(%#x) = %s, want %s", op.Byte(), got, op)
		}
	}

	if _, ok := OpFromByte(0xEE); ok {
		t.Fatal("invalid opcode resolved")
	}
}

func TestCategory(t *testing.T) {
	tests := map[Op]Category{
		LoadConst:      CategoryStack,
		LoadName:       CategoryVariable,
		RShift:         CategoryArithmetic,
		Not:            CategoryComparison,
		JumpIfTrue:     CategoryControlFlow,
		Return:         CategoryFunction,
		DictSet:        CategoryCollection,
		LoadSlice:      CategorySubscriptAttribute,
		UnpackSequence: CategoryIteration,
		ImportFrom:     CategoryModule,
		Print:          CategoryIO,
		Halt:           CategoryVMControl,
	}

	for op, want := range tests {
		got, ok := op.Category()
		if !ok {
			t.Fatalf("%s category not found", op)
		}
		if got != want {
			t.Fatalf("%s category = %s, want %s", op, got, want)
		}
	}

	if _, ok := Op(0xB0).Category(); ok {
		t.Fatal("unknown category resolved")
	}
	if _, ok := Op(0x07).Category(); ok {
		t.Fatal("undefined stack-range opcode category resolved")
	}
}

func TestOperatorMaps(t *testing.T) {
	binary := BinaryOpMap()
	if binary["+"] != Add || binary["//"] != FloorDiv || binary[">>"] != RShift {
		t.Fatalf("binary map mismatch: %#v", binary)
	}
	if len(binary) != 12 {
		t.Fatalf("binary map length = %d, want 12", len(binary))
	}
	if op, ok := BinaryOpcode("**"); !ok || op != Power {
		t.Fatalf("BinaryOpcode(**) = %s, %t", op, ok)
	}
	if _, ok := BinaryOpcode("???"); ok {
		t.Fatal("unknown binary operator resolved")
	}

	compare := CompareOpMap()
	if compare["=="] != CmpEq || compare["not in"] != CmpNotIn {
		t.Fatalf("compare map mismatch: %#v", compare)
	}
	if len(compare) != 8 {
		t.Fatalf("compare map length = %d, want 8", len(compare))
	}
	if op, ok := CompareOpcode("in"); !ok || op != CmpIn {
		t.Fatalf("CompareOpcode(in) = %s, %t", op, ok)
	}
	if _, ok := CompareOpcode("contains"); ok {
		t.Fatal("unknown compare operator resolved")
	}

	augmented := AugmentedAssignMap()
	if augmented["+="] != Add || augmented["**="] != Power {
		t.Fatalf("augmented map mismatch: %#v", augmented)
	}
	if len(augmented) != 12 {
		t.Fatalf("augmented map length = %d, want 12", len(augmented))
	}
	if op, ok := AugmentedAssignOpcode("<<="); !ok || op != LShift {
		t.Fatalf("AugmentedAssignOpcode(<<=) = %s, %t", op, ok)
	}
	if _, ok := AugmentedAssignOpcode("="); ok {
		t.Fatal("unknown augmented operator resolved")
	}

	unary := UnaryOpMap()
	if unary["-"] != Negate || unary["~"] != BitNot {
		t.Fatalf("unary map mismatch: %#v", unary)
	}
	if len(unary) != 2 {
		t.Fatalf("unary map length = %d, want 2", len(unary))
	}
	if op, ok := UnaryOpcode("~"); !ok || op != BitNot {
		t.Fatalf("UnaryOpcode(~) = %s, %t", op, ok)
	}
	if _, ok := UnaryOpcode("not"); ok {
		t.Fatal("unknown unary operator resolved")
	}
}

func TestMapsAndAllOpsReturnCopies(t *testing.T) {
	ops := AllOps()
	ops[0] = Halt
	if AllOps()[0] != LoadConst {
		t.Fatal("AllOps returned shared storage")
	}

	binary := BinaryOpMap()
	binary["custom"] = Halt
	if _, ok := BinaryOpMap()["custom"]; ok {
		t.Fatal("BinaryOpMap returned shared storage")
	}
}

func TestStringUnknown(t *testing.T) {
	if got := Op(0xB0).String(); got != "Op(unknown)" {
		t.Fatalf("unknown string = %q", got)
	}
}

File: code/packages/elixir/starlark_compiler/test/starlark_compiler_test.exs
defmodule CodingAdventures.StarlarkCompilerTest do
  use ExUnit.Case, async: true

  alias CodingAdventures.StarlarkCompiler

  test "keeps Rust-compatible byte values" do
    assert StarlarkCompiler.op_byte(:load_const) == 0x01
    assert StarlarkCompiler.op_byte(:add) == 0x20
    assert StarlarkCompiler.op_byte(:cmp_eq) == 0x30
    assert StarlarkCompiler.op_byte(:jump) == 0x40
    assert StarlarkCompiler.op_byte(:make_function) == 0x50
    assert StarlarkCompiler.op_byte(:build_list) == 0x60
    assert StarlarkCompiler.op_byte(:load_subscript) == 0x70
    assert StarlarkCompiler.op_byte(:get_iter) == 0x80
    assert StarlarkCompiler.op_byte(:load_module) == 0x90
    assert StarlarkCompiler.op_byte(:print) == 0xA0
    assert StarlarkCompiler.op_byte(:halt) == 0xFF
    assert StarlarkCompiler.op_byte(:missing) == nil
  end

  test "round trips every defined opcode" do
    for op <- StarlarkCompiler.all_ops() do
      assert StarlarkCompiler.op_from_byte(StarlarkCompiler.op_byte(op)) == op
    end

    assert StarlarkCompiler.op_from_byte(0xEE) == nil
  end

  test "classifies opcodes by high nibble" do
    assert StarlarkCompiler.op_category(:load_const) == :stack
    assert StarlarkCompiler.op_category(:load_name) == :variable
    assert StarlarkCompiler.op_category(:r_shift) == :arithmetic
    assert StarlarkCompiler.op_category(:not) == :comparison
    assert StarlarkCompiler.op_category(:jump_if_true) == :control_flow
    assert StarlarkCompiler.op_category(:return) == :function
    assert StarlarkCompiler.op_category(:dict_set) == :collection
    assert StarlarkCompiler.op_category(:load_slice) == :subscript_attribute
    assert StarlarkCompiler.op_category(:unpack_sequence) == :iteration
    assert StarlarkCompiler.op_category(:import_from) == :module
    assert StarlarkCompiler.op_category(:print) == :io
    assert StarlarkCompiler.op_category(:halt) == :vm_control
    assert StarlarkCompiler.op_category(:missing) == nil
  end

  test "maps binary operators" do
    map = StarlarkCompiler.binary_op_map()

    assert map["+"] == :add
    assert map["-"] == :sub
    assert map["*"] == :mul
    assert map["/"] == :div
    assert map["//"] == :floor_div
    assert map["%"] == :mod
    assert map["**"] == :power
    assert map["&"] == :bit_and
    assert map["|"] == :bit_or
    assert map["^"] == :bit_xor
    assert map["<<"] == :l_shift
    assert map[">>"] == :r_shift
    assert StarlarkCompiler.binary_opcode("+") == :add
    assert StarlarkCompiler.binary_opcode("???") == nil
    assert map_size(map) == 12
  end

  test "maps comparison operators" do
    map = StarlarkCompiler.compare_op_map()

    assert map["=="] == :cmp_eq
    assert map["!="] == :cmp_ne
    assert map["<"] == :cmp_lt
    assert map[">"] == :cmp_gt
    assert map["<="] == :cmp_le
    assert map[">="] == :cmp_ge
    assert map["in"] == :cmp_in
    assert map["not in"] == :cmp_not_in
    assert StarlarkCompiler.compare_opcode("in") == :cmp_in
    assert StarlarkCompiler.compare_opcode("contains") == nil
    assert map_size(map) == 8
  end

  test "maps augmented assignment and unary operators" do
    augmented = StarlarkCompiler.augmented_assign_map()

    assert augmented["+="] == :add
    assert augmented["-="] == :sub
    assert augmented["*="] == :mul
    assert augmented["/="] == :div
    assert augmented["//="] == :floor_div
    assert augmented["%="] == :mod
    assert augmented["&="] == :bit_and
    assert augmented["|="] == :bit_or
    assert augmented["^="] == :bit_xor
    assert augmented["<<="] == :l_shift
    assert augmented[">>="] == :r_shift
    assert augmented["**="] == :power
    assert StarlarkCompiler.augmented_assign_opcode("**=") == :power
    assert StarlarkCompiler.augmented_assign_opcode("=") == nil
    assert map_size(augmented) == 12

    unary = StarlarkCompiler.unary_op_map()
    assert unary["-"] == :negate
    assert unary["~"] == :bit_not
    assert StarlarkCompiler.unary_opcode("~") == :bit_not
    assert StarlarkCompiler.unary_opcode("not") == nil
    assert map_size(unary) == 2
  end
end
